File: app/src/main/java/com/home/young/myPassword/application/AES.java
package com.home.young.myPassword.application;

import android.annotation.SuppressLint;
import android.util.Base64;

import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.SecureRandom;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.KeyGenerator;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

public class AES {
    private static final String AES = "AES";
    private static final int KEYSIZE = 128;

    public static String encrypt(String cleartext, String seed) throws NoSuchProviderException,
            NoSuchAlgorithmException,
            IllegalBlockSizeException,
            InvalidKeyException,
            BadPaddingException,
            NoSuchPaddingException {
        byte[] rawKey = getRawKey(seed.getBytes());
        byte[] result = encrypt(rawKey, cleartext.getBytes());
        return Base64.encodeToString(result, 0);
    }

    public static String decrypt(String encrypted, String seed) throws NoSuchProviderException,
            NoSuchAlgorithmException,
            IllegalBlockSizeException,
            InvalidKeyException,
            BadPaddingException,
            NoSuchPaddingException {
        byte[] rawKey = getRawKey(seed.getBytes());
        byte[] enc = Base64.decode(encrypted, 0);
        byte[] result = decrypt(rawKey, enc);
        return new String(result);
    }

    @SuppressLint("TrulyRandom")
    private static byte[] getRawKey(byte[] seed) throws NoSuchAlgorithmException,
            NoSuchProviderException {
        KeyGenerator keyGenerator = KeyGenerator.getInstance(AES);
        SecureRandom sr = SecureRandom.getInstance("SHA1PRNG", "Crypto");
        sr.setSeed(seed);
        keyGenerator.init(KEYSIZE, sr);
        SecretKey secretKey = keyGenerator.generateKey();
        byte[] raw = secretKey.getEncoded();
        return raw;
    }

    private static byte[] encrypt(byte[] raw, byte[] clear) throws NoSuchAlgorithmException,
            NoSuchPaddingException,
            InvalidKeyException,
            IllegalBlockSizeException,
            BadPaddingException {
        SecretKeySpec secretKeySpec = new SecretKeySpec(raw, AES);
        Cipher cipher = Cipher.getInstance(AES);
        cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec);
        byte[] encrypted = cipher.doFinal(clear);
        return encrypted;
    }

    private static byte[] decrypt(byte[] raw, byte[] encrypted) throws NoSuchPaddingException,
            NoSuchAlgorithmException,
            InvalidKeyException,
            BadPaddingException,
            IllegalBlockSizeException {
        SecretKeySpec secretKeySpec = new SecretKeySpec(raw, AES);
        Cipher cipher = Cipher.getInstance(AES);
        cipher.init(Cipher.DECRYPT_MODE, secretKeySpec);
        byte[] decrypted = cipher.doFinal(encrypted);
        return decrypted;
    }
}

File: app/src/main/kotlin/org/tianjyan/pwd/application/AES.kt
package org.tianjyan.pwd.application

import android.annotation.SuppressLint
import android.util.Base64
import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import java.security.NoSuchProviderException
import java.security.SecureRandom
import javax.crypto.*
import javax.crypto.spec.SecretKeySpec

object AES {
    private val AES = "AES"
    private val KEYSIZE = 128

    @Throws(NoSuchProviderException::class, NoSuchAlgorithmException::class, IllegalBlockSizeException::class, InvalidKeyException::class, BadPaddingException::class, NoSuchPaddingException::class)
    fun encrypt(cleartext: String, seed: String): String {
        val rawKey = getRawKey(seed.toByteArray())
        val result = encrypt(rawKey, cleartext.toByteArray())
        return Base64.encodeToString(result, 0)
    }

    @Throws(NoSuchProviderException::class, NoSuchAlgorithmException::class, IllegalBlockSizeException::class, InvalidKeyException::class, BadPaddingException::class, NoSuchPaddingException::class)
    fun decrypt(encrypted: String, seed: String): String {
        val rawKey = getRawKey(seed.toByteArray())
        val enc = Base64.decode(encrypted, 0)
        val result = decrypt(rawKey, enc)
        return String(result)
    }

    @SuppressLint("TrulyRandom")
    @Throws(NoSuchAlgorithmException::class, NoSuchProviderException::class)
    private fun getRawKey(seed: ByteArray): ByteArray {
        val keyGenerator = KeyGenerator.getInstance(AES)
        val sr = SecureRandom.getInstance("SHA1PRNG", "Crypto")
        sr.setSeed(seed)
        keyGenerator.init(KEYSIZE, sr)
        val secretKey = keyGenerator.generateKey()
        return secretKey.encoded
    }

    @Throws(NoSuchAlgorithmException::class, NoSuchPaddingException::class, InvalidKeyException::class, IllegalBlockSizeException::class, BadPaddingException::class)
    private fun encrypt(raw: ByteArray, clear: ByteArray): ByteArray {
        val secretKeySpec = SecretKeySpec(raw, AES)
        val cipher = Cipher.getInstance(AES)
        cipher.init(Cipher.ENCRYPT_MODE, secretKeySpec)
        return cipher.doFinal(clear)
    }

    @Throws(NoSuchPaddingException::class, NoSuchAlgorithmException::class, InvalidKeyException::class, BadPaddingException::class, IllegalBlockSizeException::class)
    private fun decrypt(raw: ByteArray, encrypted: ByteArray): ByteArray {
        val secretKeySpec = SecretKeySpec(raw, AES)
        val cipher = Cipher.getInstance(AES)
        cipher.init(Cipher.DECRYPT_MODE, secretKeySpec)
        return cipher.doFinal(encrypted)
    }
}